xformers/components/attention/blocksparse.py

```diff
@@ -86,39 +85,24 @@ def __init__(
             ), "Only block sizes in [16, 32, 64] are supported"
 
             super().__init__()
+
+            self.causal = causal
+
             self.attn_drop = torch.nn.Dropout(dropout, inplace=False)
 
             # Pure blocksparse data
             self.layout = layout
             self.block_size = block_size
 
-            # blocksparse operators
-            self.sparse_dot_sdd = blocksparse_matmul(
-                self.layout,
-                self.block_size,
-                "sdd",
-                trans_a=False,
-                trans_b=True,
-            )
-            self.sparse_dot_dsd = blocksparse_matmul(
-                self.layout,
-                self.block_size,
-                "dsd",
-                trans_a=False,
-                trans_b=False,
-            )
-            self.sparse_softmax = blocksparse_softmax(self.layout, self.block_size)
-
             # make sure that the head dimension is not folded down with the batch
             self.requires_head_dimension = True
 
             # key padding mask and attention mask must be passed in separately
-            self.requires_separate_masks = True
             self.requires_same_k_q_dimensions = True
 
-            # Properties specific to this attention mechanism
-            self.supports_attention_mask = True
-            self.supports_key_padding_mask = True
+            # The underlying triton op does not support per element attention mask
```

Comment:
see previous PR which introduced these flags, we can now flip them in this case

Reply:
This is in principle a BC-breaking change. Should we bother, or should we just follow what Triton does?

Reply:
basically we don't have much choice, short of implementing blocksparse ourselves ? Phil arguments were that it was not typically used (short of causal, which is supported), and that the attention mask (additive -> floats) ended up taking a significant amount of space in memory. My take would be that we have a in-house fallback since people can use sparse attention, and else they can stick to the current pip release for some time, so not something that I would do every day (breaking BC), but in that case that was ok ?